Otus-DevOps-2018-02/ifqthenp_microservices
# HW 15. Docker images & microservices (docker-3-alpine)

# Completed Homework №15

:large_orange_diamond: [x] Extra tasks for self-study

_For main homework solution please refer to `docker-3` branch_

## The following tasks have been completed

- Extra task for self-study to create Docker images based on Linux Alpine and to reduce image size even more using [Dockerfile best practices](https://docs.docker.com/develop/develop-images/dockerfile_best-practices/)

## How to get project running

1. Connect to the previously created `docker-host` on GCP
    - `eval $(docker-machine env docker-host)`
2. `cd` into `src` folder and build Docker images
    - `docker pull mongo:latest`
    - `docker build -t ifqthenp/otus-reddit/post:2.0-alpine ./post-py`
    - `docker build -t ifqthenp/otus-reddit/comment:2.0-alpine ./comment`
    - `docker build -t ifqthenp/otus-reddit/ui:2.0-alpine ./ui`
3. Create Docker network for Reddit application
    - `docker network create reddit`
4. Create volume for MongoDB
    - `docker volume create reddit_db`
5. Run containers 
    - `docker run -d --network=reddit --network-alias=post_db --network-alias=comment_db -v reddit_db:/data/db mongo:latest`
    - `docker run -d --network=reddit --network-alias=post ifqthenp/otus-reddit/post:2.0-alpine`
    - `docker run -d --network=reddit --network-alias=comment ifqthenp/otus-reddit/comment:2.0-alpine`
    - `docker run -d --network=reddit -p 9292:9292 ifqthenp/otus-reddit/ui:2.0-alpine`
6. Kill running containers
    - `docker kill $(docker ps -q)`

## How to check that project actually works

- Ensure that the build passed TravisCI tests
  - [#andrei_bogomja](https://devops-team-otus.slack.com/messages/C9M6L1XB3) Slack channel
  - actual GitHub pull request page
- Run `docker images` command and compare the sizes of `2.0-alpine` tagged images with previously created images
- Type URL in browser address bar `http://35.197.227.48:9292/` and post a link or leave a comment
- Kill running containers and relaunch them again to make sure that the data has not been lost

## PR checklist
